### PR TITLE
contracts: Run start function

### LINKED ISSUE #116: Contracts: change `ensure_no_start` to `start`
Blocked by https://github.com/paritytech/substrate/issues/13639

### Is there an existing issue?

- [X] I have searched the existing issues

### Experiencing problems? Have you tried our Stack Exchange first?

- [X] This is not a support question.

### Motivation

Some languages must init basic env(GC) for later call.
It's very hard to remove `start` function.

BTW, It's a break change since `sp-sandbox` support to call `start` before other call. Now not support `start` anymore.

> Reject contracts with start functions
Having a start function in contracts give us all kinds of headaches. They were allowed but neither Rust nor solang make use of them. Strictly speaking this is a change that breaks existing contracts with start functions but that would be some very exotic contracts written in assembly and we prefer to get rid of that before a substantial amount of contracts start depending on it. We might need to roll it back when too many people start complaining.

### Request

Support to call `start` before `deploy`/`call`

### Solution

Also compute gas cost for `start` function.

### Are you willing to help with this request?

Maybe (please elaborate above)

### PR DESCRIPTION
**Pull request by athei**
Fixes #116 

Start function wasn't allowed in a contract. Now it is allowed and is being run.

It was disallowed because it is not used by Rust and supporting it made the code more complex. However, not running the start function violates the wasm standard. This makes life harder for some languages (see linked ticket).

### CODE REVIEW

File: substrate/frame/contracts/src/wasm/mod.rs [hunk 1]
```diff
@@ -1878,32 +1885,37 @@ mod tests {
...
-	fn start_fn_illegal() {
-		let output = execute(START_FN_ILLEGAL, vec![], MockExt::default());
-		assert_err!(output, <Error<Test>>::CodeRejected,);
+	fn start_fn_does_run() {
+		let mut ext = MockExt::default();
+		execute(START_FN_DOES_RUN, vec![], &mut ext).unwrap();
+		assert_eq!(
+			ext.events[0].1,
+			[0x00_u8, 0x01, 0x2a, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0xe5, 0x14, 0x00]
+		);
```
Comment (pgherveou) [inline substrate/frame/contracts/src/wasm/mod.rs:1918]: start runs for both call and deploy? maybe we check that in the assert too

**Review by pgherveou — APPROVED** [review]

**Comment by pgherveou** [comment]
> Fixes #116
> 
> Start function wasn't allowed in a contract. Now it is allowed and is being run.

could we just tell quickly why is was disallowed and why do we allow it now in the PR description?

File: substrate/frame/contracts/src/wasm/mod.rs (hunk 1 above)
Comment (xermicus) [reply substrate/frame/contracts/src/wasm/mod.rs:1918]: It should really be independent of the called export, i.e. should work for any export:

```rust
		match instance.start(&mut store) {
			Ok(instance) => {
				let exported_func = instance
					.get_export(&store, function.identifier())
					.and_then(|export| export.into_func())
					.ok_or_else(|| {
						log::error!(target: LOG_TARGET, "failed to find entry point");
						Error::<T>::CodeRejected
					})?;

				// ...
			},
			Err(err) => process_result(store, Err(err)),
		}
```

However I agree, the same testcase for `deploy` wouldn't hurt :)

File: substrate/frame/contracts/src/wasm/mod.rs [hunk 2]
```diff
@@ -410,25 +408,34 @@ impl<T: Config> Executable<T> for WasmBlob<T> {
...
+			Ok(instance) => {
+				let exported_func = instance
+					.get_export(&store, function.identifier())
+					.and_then(|export| export.into_func())
+					.ok_or_else(|| {
+						log::error!(target: LOG_TARGET, "failed to find entry point");
+						Error::<T>::CodeRejected
+					})?;
+
+				if let &ExportedFunction::Constructor = function {
```
Comment (xermicus) [inline substrate/frame/contracts/src/wasm/mod.rs:430]: What if the `start` function internally calls the `deploy` export, for `call` (and vice versa)? I suppose nothing bad should happen as the contract code is completely untrusted right?

File: substrate/frame/contracts/src/wasm/mod.rs (hunk 1 above)
Comment (athei) [reply substrate/frame/contracts/src/wasm/mod.rs:1918]: Yes the start function always runs once after a wasm instance is created. Will add a test.

File: substrate/frame/contracts/src/wasm/mod.rs (hunk 2 above)
Comment (athei) [reply substrate/frame/contracts/src/wasm/mod.rs:430]: This is possible and there is nothing wrong with it. The start function is metered and doesn't have any special semantics.

Comment (athei) [reply substrate/frame/contracts/src/wasm/mod.rs:430]: Thinking about it again: If the contract traps or terminates etc. in the start function we should do no further call into the blob. Needs some thinking rework.

**Comment by athei** [comment]
Do not merge, yet.

File: substrate/frame/contracts/src/wasm/mod.rs (hunk 1 above)
Comment (athei) [reply substrate/frame/contracts/src/wasm/mod.rs:1904]: Added the test.

**Comment by athei** [comment]
> > Fixes #116
> > Start function wasn't allowed in a contract. Now it is allowed and is being run.
> 
> could we just tell quickly why is was disallowed and why do we allow it now in the PR description?

Done

**Review by pgherveou — APPROVED** [review]

**Pull request merged**